General SIP-14 Future method implementation cleanup

viktorklang opened the pull request:
Cleaning up implementations of Future methods to try to keep method size down to allow for easier JIT-ing (smaller method sizes) and increasing code reuse.

Review by: @adriaanm & @phaller

adriaanm commented:
suggested commit message:

"Faster Future: reduce byte code of its methods

[explanation about JIT-ing etc goes here]
"

retronym commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -116,9 +116,9 @@ trait Future[+T] extends Awaitable[T] {
    */
   def onSuccess[U](pf: PartialFunction[T, U])(implicit executor: ExecutionContext): Unit = onComplete {
-    case Success(v) if pf isDefinedAt v => pf(v)
+    case Success(v) => pf.applyOrElse(v, (_: T) => null.asInstanceOf[U])
```
Perhaps lift this constant function closure out to a constant in the companion object? It will probably need a BC exception, though, as it might end up bytecode public.

retronym replied:
Can you also please add a doc comment explaining what's going on with the `null`? Is it possible to get there? If not, should it be an assertion?

viktorklang replied:
Yeah, it's totally legal to end up in the null case.
The question though is if we should change the signature to be a Function and then check if it's partial and then applyOrElse and just straight apply if it isn't. (this would also be a non-issue with the function/partialfunction proposal I was talking about during flatmap.no)

Hoisting the literal is a good idea, what is the best way of doing it from a BC perspective?

retronym replied:
I don't like the idea of type-casing on `Function` / `PartialFunction`, I would leave the signature as is.

It would read more clearly to me to return `()` rather than `null` from this function, then no casting is needed.

``` scala
scala> def foo[A, B](pf: PartialFunction[A, B])(a: A) { pf.applyOrElse[A, Any](a, (a: A) => ()) }
foo: [A, B](pf: PartialFunction[A,B])(a: A)Unit
```

The natural home for the hoisted `private val constUnit = (a: Any) => ()` is in the companion. But we might defer that until 2.11 for forward BC.

Until then, you could abuse the hoisted closure that sits behind `Predef.conforms`:

```
def foo[A, B](pf: PartialFunction[A, B])(a: A) { pf.applyOrElse[A, Any](a, Predef.conforms[A]) }
foo: [A, B](pf: PartialFunction[A,B])(a: A)Unit
```

retronym commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -134,9 +134,9 @@ trait Future[+T] extends Awaitable[T] {
    */
   def onFailure[U](callback: PartialFunction[Throwable, U])(implicit executor: ExecutionContext): Unit = onComplete {
-    case Failure(t) if NonFatal(t) && callback.isDefinedAt(t) => callback(t)
+    case Failure(t) => callback.applyOrElse(t, (_: Throwable) => null.asInstanceOf[U])
```
ditto for the use of a constant function.

viktorklang replied:
+1

retronym commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -221,19 +217,10 @@ trait Future[+T] extends Awaitable[T] {
+    onComplete { //TODO ideally ``v => p complete v.transform(s, f)`` if Try gets ``transform``
+      case Success(r) => p complete Try(s(r))
+      case Failure(t) => p complete Try(throw f(t))
+    }
```
Code speaks louder than comments:

``` scala
// TODO move to Try for 2.11.0
def transformTry[A](v: Try[A])(f => Throwable => Throwable, s: A => B = v match {
  case Success(r) => Try(s(r))
  case Failure(t) => Try(f(t)) // will throw fatal errors!
}
onComplete { v => transformTry(v)(f, s) }
```

viktorklang replied:
But the reason I didn't do that is because the transform method on Try has the wrong signature and introducing another method with the right signature and with an inconsistent name is sort of confusing.

retronym replied:
My suggestion is to create and use a local method right here. The name isn't so important.

retronym commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -303,34 +268,14 @@ trait Future[+T] extends Awaitable[T] {
-  }
+  def filter(pred: T => Boolean)(implicit executor: ExecutionContext): Future[T] =
+    map {
+      r  => if(pred(r)) r else throw new NoSuchElementException("Future.filter predicate is not satisfied")
```
space needed after the `if` token.

viktorklang replied:
But you're OK with the 2 spaces after the r? :-)

retronym replied:
Oh that's where it got to!

retronym commented on test/files/jvm/scala-concurrent-tck.scala:
```diff
@@ -572,9 +572,9 @@ trait FutureCombinators extends TestBase {
+      case e =>
         done()
-        assert(e == cause)
+        assert(e eq cause)
```
Can you comment as to why the test changed?

viktorklang replied:
I'm testing identity and not equality to make sure that the exception is passed through unboxed/uncopied.

retronym commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -186,13 +186,12 @@ trait Future[+T] extends Awaitable[T] {
    *  and throws a corresponding exception if the original future fails.
    */
   def failed: Future[Throwable] = {
+    implicit val ec = internalExecutor
```
What behaviour did this change? Is it tested?

viktorklang replied:
the internalExecutor was de-implicited so that it's more obvious what's used and when.

adriaanm commented:
Aside from repeating my request for documenting the reasoning about the optimization, I'm not qualified to LGTM this. We're already a week past the 2.10.2 deadline, so I'm afraid this will have to wait.

viktorklang commented:
Commit message updated

JamesIry commented:
@adriaanm, @phaller, is this ready to go?

adriaanm commented:
I defer to @phaller.

phaller commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -245,19 +236,7 @@ trait Future[+T] extends Awaitable[T] {
-        }
-    }(executor)
-
+    onComplete { v => p complete { v map f } }
```
Very nice!

phaller replied:
`v map f` is a value, though, so maybe it'd be better to pass it using regular parentheses:

``` scala
    onComplete { v => p.complete(v.map(f)) }
```

phaller commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -270,20 +249,10 @@ trait Future[+T] extends Awaitable[T] {
-        }
-    }(executor)
-
+      case Success(v) => try f(v) onComplete { p.complete } catch { case NonFatal(t) => p failure t }
```
I think using curly braces (in `{ p.complete }`) when the function doesn't take parameters is a bit confusing. How about this:

``` scala
try f(v).onComplete(p.complete) catch { case NonFatal(t) => p failure t }
```

phaller commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -303,34 +272,14 @@ trait Future[+T] extends Awaitable[T] {
-  }
+  def filter(pred: T => Boolean)(implicit executor: ExecutionContext): Future[T] =
+    map {
+      r => if (pred(r)) r else throw new NoSuchElementException("Future.filter predicate is not satisfied")
```
Nice!

phaller commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -352,22 +301,10 @@ trait Future[+T] extends Awaitable[T] {
+  def collect[S](pf: PartialFunction[T, S])(implicit executor: ExecutionContext): Future[S] =
+    map {
+      r => pf.applyOrElse(r, (t: T) => throw new NoSuchElementException("Future.collect partial function is not defined at: " + t))
+    }
```
Yep

phaller commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -383,9 +320,7 @@ trait Future[+T] extends Awaitable[T] {
-
-    onComplete { case tr => p.complete(tr recover pf) }(executor)
-
+    onComplete { v => p complete { v recover pf } }
```
Again, since `v recover pf` is a value, I'd rather pass it using regular parentheses:

``` scala
    onComplete { v => p.complete(v.recover(pf)) }
```

phaller commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -427,19 +355,12 @@ trait Future[+T] extends Awaitable[T] {
-        that onFailure {
-          case f => p failure f
-        }
+      case Success(s) => that onComplete { c => p complete { c.map((s, _)) } }
```
``` scala
    that onComplete { c => p.complete(c map { (s, _) }) }
```

makes it clear what's a function (`{}`) and what isn't (`()`)

phaller commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -470,23 +392,13 @@ trait Future[+T] extends Awaitable[T] {
-
-    p.future
+    require(boxedClass ne null)
+    map(s => boxedClass.cast(s).asInstanceOf[S])
```
Much nicer!

som-snytt commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -220,20 +218,13 @@ trait Future[+T] extends Awaitable[T] {
-        }
-    }(executor)
-
+    onComplete { v => transformTry(v)(s, f) }
```
Is there a missing p complete?

viktorklang commented:
I'm closing this for a little while, I am working on the feedback and have discovered some issues in the TCK that I am addressing before reopening, thanks for all the feedback!

viktorklang commented:
Reopening this ticket. Addressed all comments AFAICT but also rewrote and fixed an arguably quite suboptimal test suite.

@phaller, @retronym please let me know what you think

viktorklang commented:
PLS SYNC

gkossakowski commented:
@phaller, @retronym: What's the status of this PR?

retronym commented:
Can you please separate the refactoring of the test suite from the refactoring of the implementation? They should arrive in separate commits, for our reviewing pleasure :)

Sorry about asking for the busy work, but our usual mild paranoia really must be amplified when we're dealing with concurrency.

viktorklang committed:
Cleaning up method implementations in Future
    Optimizations:
    1) Avoiding isDefinedAt + apply and using applyOrElse to allow for optimizations later
    2) Reducing method sizes to be more JIT + inliner friendly
    3) Reusing core combinators to reuse inliner/JIT optimizations and be more code-cache friendly

viktorklang committed:
Refactoring to the scala-concurrent-tck.scala
  - there were numerous logical issues with the former implementation
    - failed assertions may not fail the build
  - there was a lot of ceremony and noise

viktorklang commented:
Done

phaller commented:
I'll have another look as soon as I'm at a quiet place with wifi (currently on vacation where these things are not always readily available), which is probably going to be this evening.

phaller commented:
Apart from the above nitpick, LGTM

viktorklang commented:
Considering the massive amount of Boyscouting and cleanup and the extremely long time it has taken to get this PR accepted, could I have some slack for the line of code which already contains an ugly cast @phaller ?

phaller commented:
OK, sure! I agree, it's (very) fine to merge.

phaller commented:
LGTM

viktorklang commented:
Thanks :-)

viktorklang commented:
Ok, who does the honors?

phaller commented:
Can you merge, @adriaanm? Thanks!
